### PR TITLE
PnP denoiser models

### PR DESCRIPTION
**Pull request by fsherry**
Attached are some LION implementations of deep denoisers that are commonly used for plug-and-play image restoration.

### CODE REVIEW

**Comment by AnderBiguri** [comment]
Hey @fsherry Thanks a lot!

Can you also add scripts of how you train them and evaluate them? Even if they are messy, its good for me to have a reference

**Comment by fsherry** [comment]
Ok, my training scripts and evaluation scripts are in there now too :)

File: LION/models/PnP/drunet.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+            raise ValueError(
+                f'`cite_format` "{cite_format}" is not understood, only "MLA" and "bib" are supported'
+            )
+
+    def forward(self, x0, noise_level: Optional[float] = None):
+        if self.model_parameters.use_noise_level:
+            assert (
+                isinstance(noise_level, float) and noise_level >= 0.0
+            ), "`noise_level` must be a non-negative float"
+            x0 = torch.cat((x0, noise_level * torch.ones_like(x0)), dim=1)
```
Comment (AnderBiguri) [inline LION/models/PnP/drunet.py:272]: Question: si in here you concatenate a number for the noise level? Is this like a "label" for the noise?

File: LION/models/PnP/gs_drunet.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            print("year = {2022}")
+            print("}")
+        else:
+            raise ValueError(
+                f'`cite_format` "{cite_format}" is not understood, only "MLA" and "bib" are supported'
+            )
+
+    def obj_grad(self, x, noise_level: Optional[float] = None):
+        x = x.detach()
+        x.requires_grad = True
```
Comment (AnderBiguri) [inline LION/models/PnP/gs_drunet.py:63]: Can you explain this? So you remove it from the computational graph, but then you start a new computaional graph by requiring grad?

Given that `forward` itself has no loops, how is this doing anything? Or is it to do with something about the way you trained?

File: scripts/benchmarking/train_gaussian_denoiser.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+    experiment_name += f"_n_blocks={params['n_blocks']}"
+experiment_name += f"_noise_level={params['noise_level']}_{experiment_id}"
+print(experiment_name)
+print(config)
+wandb.init(project="benchmarking_ct", config=config, name=experiment_name)
+
+optimiser = torch.optim.Adam(model.parameters(), lr=params["lr"], betas=(0.9, 0.9))
+random_crop = RandomCrop((256, 256))
+random_erasing = RandomErasing()
+experiment = ct_benchmarking.GroundTruthCT()
```
Comment (AnderBiguri) [inline scripts/benchmarking/train_gaussian_denoiser.py:154]: You may have missed to add some files?

File: scripts/benchmarking/train_gaussian_denoiser.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        x = (x - x_min) / (x_max - x_min)
+        patches = random_erasing(torch.cat([random_crop(x) for _ in range(5)], dim=0))
+        optimiser.zero_grad()
+        y = patches + params["noise_level"] * torch.randn_like(patches)
+        recon = model(y)
+        loss = torch.mean((recon - patches) ** 2)
+        loss.backward()
+        grad_norm = mean_grad_norm(model)
+        losses.append(loss.item())
+        optimiser.step()
```
Comment (AnderBiguri) [inline scripts/benchmarking/train_gaussian_denoiser.py:197]: So you erase some patches, then step on the loss and then test on the patches+nose? Or `random_erasing` gives you the patches??

You did this because of memory, right?

File: LION/models/PnP/drunet.py (hunk 1 above)
Comment (fsherry) [reply LION/models/PnP/drunet.py:272]: Yes, I didn't use it in the benchmarking paper, but the paper that introduced DRUNet suggests this. Then you can train on multiple noise levels, while inputting both noisy images and the noise level, and you hope that it generalises to a spectrum of noise levels.

File: LION/models/PnP/gs_drunet.py (hunk 2 above)
Comment (fsherry) [reply LION/models/PnP/gs_drunet.py:63]: This gives you a new tensor with the same underlying data but which doesn't interfere with previous things done on `x`. If I removed line 62, I would be messing with the `x` that the user has input (undesirable), whereas now after line 62 `x` refers to a new tensor so I can do with it what I want (except modify it). I could have just as well said `y = x.detach()` in line 62 and then used `y` for the rest of it. The screenshot shows what I mean:
Screenshot 2024-07-15 at 23 21 18

File: scripts/benchmarking/train_gaussian_denoiser.py (hunk 3 above)
Comment (fsherry) [reply scripts/benchmarking/train_gaussian_denoiser.py:154]: No, I'm not sure why you think so? `RandomCrop` and `RandomErasing` are from `kornia.augmentation` and imported on line 16.

File: scripts/benchmarking/train_gaussian_denoiser.py (hunk 4 above)
Comment (fsherry) [reply scripts/benchmarking/train_gaussian_denoiser.py:197]: `random_crop` selects random square patches from the image, while `random_erasing` erases random rectangles from the patches (intended to counteract overfitting). After applying those transforms, I add noise and  pass the noisy cropped+erased patches through the denoiser to train. Training on patches was necessary at least for the gradient-step denoisers because of memory constraints.

File: LION/models/PnP/gs_drunet.py (hunk 2 above)
Comment (AnderBiguri) [reply LION/models/PnP/gs_drunet.py:63]: Sorry,  I am still confused :D. Yes, I do understand what it does in principle, but when we make ML models, we don't need to detach the input of the model, right? the computational graph understands that its the input of the model and starts the graph there. given that `obj_grad` is just used in `forward`, why is this line needed here?

File: scripts/benchmarking/train_gaussian_denoiser.py (hunk 3 above)
Comment (AnderBiguri) [reply scripts/benchmarking/train_gaussian_denoiser.py:154]: I think you introduced `ct_benchmarking.GroundTruthCT`, its not in main.

Comment (fsherry) [reply scripts/benchmarking/train_gaussian_denoiser.py:154]: Yes it is, it was in this pull request https://github.com/CambridgeCIA/LION/pull/93

File: LION/models/PnP/gs_drunet.py (hunk 2 above)
Comment (fsherry) [reply LION/models/PnP/gs_drunet.py:63]: You need to do this because we do a backward pass inside `obj_grad`. Otherwise you mess up the input `x`

File: scripts/benchmarking/train_gaussian_denoiser.py (hunk 3 above)
Comment (AnderBiguri) [reply scripts/benchmarking/train_gaussian_denoiser.py:154]: Man, I literally went to the file and contr+f it and could not find it. Not my brightest day.

File: LION/models/PnP/gs_drunet.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+            x,
+            torch.ones_like(obj),
+            create_graph=self.training,
+            retain_graph=self.training,
+        )[0]
+        return obj, grad
+
+    def forward(self, x, noise_level: Optional[float] = None):
+        _, grad = self.obj_grad(x, noise_level)
+        return x - grad
```
Comment (AnderBiguri) [inline LION/models/PnP/gs_drunet.py:76]: Ah I see, this is the one that gets messed up, `x-grad`. Makes sense now, thanks :)
